File: ScriptsToProcess/NEW BATCH TO PROCESS/MakePalette.py
"""
JSON PALETTE BUILDER
================================================================================
Converts a list of program strings and visual Grasshopper Color Swatches 
into a formatted JSON dictionary.

INPUTS:
    Programs (str)   [List Access] : Names of the architectural programs.
    Colors   (Color) [List Access] : System.Drawing.Color objects from GH Swatches.

OUTPUTS:
    JSON_Palette (str) : Formatted JSON string ready for the OOP Engine.
================================================================================
"""

import json
import System.Drawing as sd
import Grasshopper as gh

# UI Metadata
ghenv.Component.Name = "JSON Palette Builder"
ghenv.Component.NickName = "MakePalette"
ghenv.Component.Description = "Zips strings and GH Colors into a JSON dictionary."

def build_palette():
    progs = globals().get('Programs', [])
    colors = globals().get('Colors', [])
    
    if not progs or not colors:
        return "{\n}", "Awaiting Data"
        
    palette_dict = {}
    
    # Safely zip the lists (stops at the shortest list to prevent crashes)
    limit = min(len(progs), len(colors))
    
    for i in range(limit):
        p_name = str(progs[i]).strip()
        c = colors[i]
        
        # Ensure we are extracting the RGB channels correctly
        if hasattr(c, 'R') and hasattr(c, 'G') and hasattr(c, 'B'):
            palette_dict[p_name] = [int(c.R), int(c.G), int(c.B)]
            
    # Format the dictionary into a clean JSON string
    json_str = json.dumps(palette_dict, indent=2)
    
    # UI Feedback
    msg = "Mapped {} Programs".format(limit)
    if len(progs) != len(colors):
        msg += "\nWarning: List length mismatch!"
        
    return json_str, msg

# Execute
json_out, ui_msg = build_palette()

# Outputs
JSON_Palette = json_out
ghenv.Component.Message = ui_msg

File: Components/MakePalette.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using Grasshopper.Kernel;
using Newtonsoft.Json;

namespace Enzyme.Components
{
    public class MakePalette : GH_Component
    {
        public MakePalette()
          : base("JSON Palette Builder", "MakePalette",
              "Zips strings and GH Colors into a JSON dictionary.",
              Enzyme.Utils.TabInfo.TabName, "Masterplan (Beta)")
        {
        }

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                Enzyme.Utils.AutoWireHelper.WireOutputPanel(this, document, 0, 220, -11, 180, 22);
            }
        }

        protected override void RegisterInputParams(GH_Component.GH_InputParamManager pManager)
        {
            pManager.AddTextParameter("Programs", "Programs", "Names of the architectural programs.", GH_ParamAccess.list);
            pManager.AddColourParameter("Colors", "Colors", "System.Drawing.Color objects from GH Swatches.", GH_ParamAccess.list);
        }

        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
        {
            pManager.AddTextParameter("JSON_Palette", "JSON_Palette", "Formatted JSON string ready for the OOP Engine.", GH_ParamAccess.item);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            List<string> progs = new List<string>();
            List<Color> colors = new List<Color>();

            bool hasProgs = DA.GetDataList(0, progs);
            bool hasColors = DA.GetDataList(1, colors);

            if (!hasProgs || !hasColors || progs.Count == 0 || colors.Count == 0)
            {
                DA.SetData(0, "{\n}");
                Message = "Awaiting Data";
                return;
            }

            int limit = Math.Min(progs.Count, colors.Count);
            var paletteDict = new Dictionary<string, int[]>();

            for (int i = 0; i < limit; i++)
            {
                string pName = progs[i]?.Trim();
                if (string.IsNullOrEmpty(pName)) continue;
                
                Color c = colors[i];
                paletteDict[pName] = new int[] { c.R, c.G, c.B };
            }

            string jsonStr = JsonConvert.SerializeObject(paletteDict, Formatting.Indented);
            
            string msg = $"Mapped {limit} Programs";
            if (progs.Count != colors.Count)
            {
                msg += "\nWarning: List length mismatch!";
            }
            
            Message = msg;
            
            DA.SetData(0, jsonStr);
        }

        protected override Bitmap Icon
        {
            get
            {
                return IconLoader.Load("MakePalette.png");
            }
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("d67c9c22-b43a-4f51-b062-8be2f42e2098"); }
        }
    }
}
